lib/bolt_server/transport_app.rb

```diff
@@ -142,6 +145,101 @@ def run_task(target, body)
       end
     end
 
+    def run_task(target, body)
+      validate_schema(@schemas["action-run_task"], body)
+
+      task_data = body['task']
+      task = Bolt::Task::PuppetServer.new(task_data['name'], task_data['metadata'], task_data['files'], @file_cache)
+      task_helper(target, task, body['parameters'] || {})
+    end
+
+    def extract_install_task(target)
+      unless target.plugin_hooks['puppet_library']['task']
+        raise BoltServer::RequestError,
+              "Target must have 'task' plugin hook"
+      end
+      install_task = target.plugin_hooks['puppet_library']['task'].split('::', 2)
+      install_task << 'init' if install_task.count == 1
+      install_task
+    end
+
+    # This helper is responsible for computing or retrieving from the cache a plugin tarball. There are
+    # two supported plugin types 'fact_plugins', and 'all_plugins'. Note that this is cached based on
+    # versioned_project as there are no plugins in the "builtin content" directory
+    def plugin_tarball(versioned_project, tarball_type)
+      tarball_types = %w[fact_plugins all_plugins]
+      unless tarball_types.include?(tarball_type)
+        raise ArgumentError,
+              "tarball_type must be one of: #{tarball_types.join(', ')}"
+      end
+      # lock this so that in the case an apply/apply_prep with multiple targets hits this endpoint
+      # the tarball computation only happens once (all the other targets will just need to read the cached data)
+      @plugin_mutex.synchronize do
+        if (tarball = @file_cache.get_cached_project_file(versioned_project, tarball_type))
+          tarball
+        else
+          new_tarball = build_project_plugins_tarball(versioned_project) do |mod|
+            search_dirs = []
+            search_dirs << mod.plugins if mod.plugins?
+            search_dirs << mod.pluginfacts if mod.pluginfacts?
+            if tarball_type == 'all_plugins'
+              search_dirs << mod.files if mod.files?
+              type_files = "#{mod.path}/types"
+              search_dirs << type_files if File.exist?(type_files)
+            end
+            search_dirs
+          end
+          @file_cache.cache_project_file(versioned_project, tarball_type, new_tarball)
+          new_tarball
+        end
+      end
+    end
+
+    # This helper is responsible for computing or retrieving task metadata for a project.
+    # It expects task name in segments and uses the combination of task name and versioned_project
+    # as a unique identifier for caching in addition to the job_id. The job id is added to protect against
+    # a case where the buildtin content is update (where the apply_helpers would be managed)
+    def project_task_metadata(versioned_project, task_name_segments, job_id)
+      cached_file_name = "#{task_name_segments.join('_')}_#{job_id}"
```

Comment:
If we're including job ID as part of the name here, then why not just cache everything under https://github.com/puppetlabs/bolt/blob/main/lib/bolt_server/transport_app.rb#L168 ? Thinking of something like

```
<job_id>/fact_plugins_tarball
<job_id>/all_plugins_tarball
<job_id>/<task_name>_metadata
...
```

(We'll need to change upload_file's caching to something like <job_id>/upload_file_cache/<....> to avoid conflicts where an uploaded file could have the same name as one of the cached components)

I'm OK with deferring that to a subsequent PR though, just thought I would bring it up here.

Reply:
Sure, For this iteration i wanted to keep per versioned project for plugin tarballs (as nothing we are shipping in bundled content needs to go in there)  I only wanted per versioned project plus job id for task data (as the apply_helpers are distributed there). I'm still trying to wrap my head around how practically caching will even work in kubernetes. Like is there even a scenario where there would be a deploy that includes bundled content that *also* would preserve the cache?

Reply:
Not sure. But worth documenting this discussion somewhere.